### PR TITLE
ADC: Rework ADC shell to remove DT_ADC_{0..2}_NAME usage.

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Comment by galak** [comment]
I had a commit that cleaned up dts_fixup.h as well, but dropping that for now because I think it causes CI to timeout.

File: drivers/adc/adc_shell.c [hunk 1]
```diff
@@ -10,6 +10,26 @@
...
+#elif DT_HAS_COMPAT(nordic_nrf_saadc)
+#define DT_DRV_COMPAT nordic_nrf_saadc
+#elif DT_HAS_COMPAT(nxp_kinetis_adc12)
+#define DT_DRV_COMPAT nxp_kinetis_adc12
+#elif DT_HAS_COMPAT(nxp_kinetis_adc16)
+#define DT_DRV_COMPAT nxp_kinetis_adc16
+#elif DT_HAS_COMPAT(st_stm32_adc)
+#define DT_DRV_COMPAT st_stm32_adc
+#else
+#error No know devicetree compatiable match for ADC shell
```
Comment (MaureenHelm) [inline drivers/adc/adc_shell.c:30]: typo: "compatiable"

**Review by MaureenHelm — APPROVED** [review]

File: drivers/adc/adc_shell.c (hunk 1 above)
Comment (galak) [reply drivers/adc/adc_shell.c:30]: fixed

**Comment by galak** [comment]
dropped the Kconfig cleanup as well for CI purposes.

**Review by nandojve — APPROVED** [review]

File: drivers/adc/adc_shell.c [hunk 2]
```diff
@@ -10,6 +10,26 @@
...
+#elif DT_HAS_COMPAT(nxp_kinetis_adc12)
+#define DT_DRV_COMPAT nxp_kinetis_adc12
+#elif DT_HAS_COMPAT(nxp_kinetis_adc16)
+#define DT_DRV_COMPAT nxp_kinetis_adc16
+#elif DT_HAS_COMPAT(st_stm32_adc)
+#define DT_DRV_COMPAT st_stm32_adc
+#else
+#error No know devicetree compatible match for ADC shell
+#endif
+
```
Comment (henrikbrixandersen) [inline drivers/adc/adc_shell.c:32]: This will prevent the ADC shell from working with multiple ADC drivers at once (e.g. SoC ADC and external, SPI ADC).

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please reconsider the solution to this to include support for multiple, different ADCs at once.

File: drivers/adc/adc_shell.c (hunk 2 above)
Comment (galak) [reply drivers/adc/adc_shell.c:32]: That limitation exist today.  There is no support for ADC shell on non-SoC ADC currently.  I'd ask we look at doing that as a follow up to this change.

Comment (henrikbrixandersen) [reply drivers/adc/adc_shell.c:32]: Fair enough.

**Review by henrikbrixandersen — APPROVED** [review]

File: drivers/adc/adc_shell.c [hunk 3]
```diff
@@ -10,6 +10,26 @@
...
+#elif DT_HAS_COMPAT(nordic_nrf_saadc)
+#define DT_DRV_COMPAT nordic_nrf_saadc
+#elif DT_HAS_COMPAT(nxp_kinetis_adc12)
+#define DT_DRV_COMPAT nxp_kinetis_adc12
+#elif DT_HAS_COMPAT(nxp_kinetis_adc16)
+#define DT_DRV_COMPAT nxp_kinetis_adc16
+#elif DT_HAS_COMPAT(st_stm32_adc)
+#define DT_DRV_COMPAT st_stm32_adc
+#else
+#error No know devicetree compatible match for ADC shell
```
Comment (nwsetec) [inline drivers/adc/adc_shell.c:30]: known

Comment (galak) [reply drivers/adc/adc_shell.c]: Fixed

**Comment by avigreen1978** [comment]
The new implenentation assumes that the first instance (inst=0) is tied to label ADC_0, which doesn't exist in stm32 for example.
For example, if you use a nucleo_l432kc board, change only adc1 status to okay in dts, then allow ADC and ADC_SHELL configs, the shell subcommand which is generated is for ADC_0, which doesn't exist.
To my understanding, instance number isn't supposed to be tied with peripheral number.
The number which should be sent to the ADC_SHELL_COMMAND macro is the peripheral number, not the instance number.

**Comment by galak** [comment]
> The new implementation assumes that the first instance (inst=0) is tied to label ADC_0, which doesn't exist in stm32 for example.
> For example, if you use a nucleo_l432kc board, change only adc1 status to okay in dts, then allow ADC and ADC_SHELL configs, the shell subcommand which is generated is for ADC_0, which doesn't exist.
> To my understanding, instance number isn't supposed to be tied with peripheral number.
> The number which should be sent to the ADC_SHELL_COMMAND macro is the peripheral number, not the instance number.

The instance 0 isn't tied to ADC_0, its just the first enabled instance.  So on STM32 this shoudl still work.

**Comment by siddh7007** [comment]
I am having similar issue on STM32L4 device, I am using disco_l475_iot1 board, I have enabled adc1 node from devicetree. Now, when I open up shell and enter "adc" is shows ADC_0 as a subcommand.
 "adc ADC_0 print" returns "Device not in device list".
 As avigreen1978 mentioned this device shows ADC_1 upon issuing"device list" command not ADC_0.
 If I try "adc ADC_1 print" it doesn't print anything in shell but only shows ADC_0.


> > The new implementation assumes that the first instance (inst=0) is tied to label ADC_0, which doesn't exist in stm32 for example.
> > For example, if you use a nucleo_l432kc board, change only adc1 status to okay in dts, then allow ADC and ADC_SHELL configs, the shell subcommand which is generated is for ADC_0, which doesn't exist.
> > To my understanding, instance number isn't supposed to be tied with peripheral number.
> > The number which should be sent to the ADC_SHELL_COMMAND macro is the peripheral number, not the instance number.
> 
> The instance 0 isn't tied to ADC_0, its just the first enabled instance. So on STM32 this shoudl still work.

**Comment by avigreen1978** [comment]
Yes, as @siddh7007 mentioned this doesn't work on STM32, but the problem can occur in many other configurations and architectures.

The root cause for this bug is the ambiguous use of the sub command name, here "ADC_0", and the device label configured in the dts.
if "ADC_0" means "the first enabled ADC instance" in line 45: `.device_name = DT_INST_LABEL(inst),` sets the device_name field in the hdl struct as "ADC_1", the device label string defined in the dts.
But, in line 111 in function get_adc_from_list(char *name): `if (!strcmp(name, adc_list[adc_idx].device_name)) {` it is compared to the argv parameter given, which is "ADC_0". thus always failing the search. 

**Note:** using "ADC_1" as shell parameter fails the shell core as this is not a valid subcommand.

This can happen in the following cases, which causes the enabled device under instance index x to no be labelled "ADC_x":
1. The arch dts device label defines are not 0-based (like in STM32 which start at ADC_1)
2. The first enabled device is not ADC_0, but ADC_1, or 2
3. The arch dts doesn't even use the ADC_x naming for its labels. it can be "A2D_0" (for digital-to-analog) or something like that.

Solution:
I suggest avoiding the double use of ADC_x for both the subcommand syntax and the device label. using index only for the subcommand. using `adc 0 read 15` for example.

If this "ADC_x" naming convention is necessary, using "ADC_0" for "first enabled instance" is wrong, and it should be "ADC_1" for the device label as defined in the dts. can be achieved by dynamically creating the subcommand syntax.

For now, this code is malfunctioning in every ST core we used.

**Comment by avigreen1978** [comment]
#27350 
@galak

**Pull request merged**